configfile: "config.yaml"

ASSAY = config["assay"]
USE_GRAPH = config.get("use_graph", False)

if ASSAY == "rna-seq":
    if USE_GRAPH:
        include: "workflow/vg_rules.smk"
    else:
        include: "workflow/rnaseq_rules.smk"
elif ASSAY == "microarray":
    include: "workflow/microarray_rules.smk"
else:
    raise ValueError(f"Unknown assay type: {ASSAY}")

rna_seq_outputs = [
    "results/clean_R1.fastq",
    "results/fastqc/clean_R1_fastqc.zip",
    "results/centrifuge/clean_R1_report.txt",
    "results/centrifuge/clean_R1_output.tsv",
    "results/fastp.json",
    "results/multiqc_report.html"
]

if config.get("r2"):
    rna_seq_outputs.extend([
        "results/fastqc/clean_R2_fastqc.zip",
        "results/centrifuge/clean_PE_report.txt",
        "results/centrifuge/clean_PE_output.tsv"
    ])

if not USE_GRAPH:
    rna_seq_outputs.extend([
        "results/salmon/quant.sf",
        "results/counts_salmon_orthogroups.tsv",
    ])

if USE_GRAPH:
    rna_seq_outputs.extend([
        "results/aligned.gam",
        "results/aligned.bam",
        "results/vg/giraffe.stats.txt",
    ])

rule all:
    input:
        rna_seq_outputs

